src/core/QCheck2.ml

```diff
@@ -0,0 +1,1868 @@
+        pp_close_box ppf ())
+      else if Seq.is_empty xs then pp_close_box ppf ()
+      else
+        (pp_print_break ppf 0 0;
+         pp_print_list
+           ~pp_sep:(fun ppf () -> pp_print_string ppf ","; pp_print_space ppf ())
+           (pp ?depth:(Option.map pred depth) inner_pp)
+           ppf
+           (List.of_seq xs);
+         pp_close_box ppf ();
+         pp_print_break ppf 0 0);
+
+      pp_print_string ppf ")"
+
+    in
+    wrapper_box inner
+
+  let rec map (f : 'a -> 'b) (a : 'a t) : 'b t =
+    let Tree (x, xs) = a in
+    let y = f x in
+    let ys = Seq.map (fun smaller_x -> map f smaller_x) xs in
+    Tree (y, ys)
+
+  (** Note that parameter order is reversed. *)
+  let (>|=) a f = map f a
+
+  let rec ap (f : ('a -> 'b) t) (a : 'a t) : 'b t =
+    let Tree (x0, xs) = a in
+    let Tree (f0, fs) = f in
+    let y = f0 x0 in
+    let ys = Seq.append (Seq.map (fun f' -> ap f' a) fs) (Seq.map (fun x' -> ap f x') xs) in
+    Tree (y, ys)
+
+  let (<*>) = ap
+
+  let liftA2 (f : 'a -> 'b -> 'c) (a : 'a t) (b : 'b t) : 'c t =
+    (a >|= f) <*> b
+
+  let rec bind (a : 'a t) (f : 'a -> 'b t) : 'b t =
+    let Tree (x, xs) = a in
+    let Tree (y, ys_of_x) = f x in
+    let ys_of_xs = Seq.map (fun smaller_x -> bind smaller_x f) xs in
+    let ys = Seq.append ys_of_xs ys_of_x in
+    Tree (y, ys)
+
+  let (>>=) = bind
+
+  let pure x = Tree (x, Seq.empty)
+
+  let rec make_primitive (shrink : 'a -> 'a Seq.t) (x : 'a) : 'a t =
+    let shrink_trees = shrink x |> Seq.map (make_primitive shrink) in
+    Tree (x, shrink_trees)
+
+  let rec opt (a : 'a t) : 'a option t =
+    let Tree (x, xs) = a in
+    let shrinks = Seq.cons (pure None) (Seq.map opt xs) in
+    Tree (Some x, shrinks)
+
+  let rec add_shrink_invariant (p : 'a -> bool) (a : 'a t) : 'a t =
+    let Tree (x, xs) = a in
+    let xs' = Seq.filter_map (fun (Tree (x', _) as t) -> if p x' then Some (add_shrink_invariant p t) else None) xs in
+    Tree (x, xs')
+end
+
+module Gen = struct
+
+  (* Maybe the type should be changed to [RS.t -> (unit -> 'a) tree] or use laziness? To avoid eagerly evaluating shrinks during generation (see implem of [(>>=)] that requires evaluating to append Sequences). *)
+  type 'a t = RS.t -> 'a Tree.t
+
+  type 'a sized = int -> Random.State.t -> 'a Tree.t
+
+  let map f x st = Tree.map f (x st)
+
+  (** Note that parameter order is reversed. *)
+  let (>|=) x f = map f x
+
+  let (<$>) = map
+
+  let pure (a : 'a) : 'a t = fun _ -> Tree.pure a
+
+  let ap (f : ('a -> 'b) t) (x : 'a t) : 'b t = fun st -> Tree.ap (f st)  (x st)
+
+  let (<*>) = ap
+
+  let liftA2 (f : 'a -> 'b -> 'c) (a : 'a t) (b : 'b t) : 'c t =
+    (a >|= f) <*> b
+
+  let liftA3 (f : 'a -> 'b -> 'c -> 'd) (a : 'a t) (b : 'b t) (c : 'c t) : 'd t =
+    (a >|= f) <*> b <*> c
+
+  let map2 = liftA2
+
+  let map3 = liftA3
+
+  let return = pure
+
+  let bind (gen : 'a t) (f : 'a -> ('b t)) : 'b t = fun st -> Tree.bind (gen st) (fun a -> f a st)
```

Comment:
* wouldn't this implementation pass a new `st0` copy (i.e. a new, independent state) to both the generator `gen` and to `f a`? I am under the impression that `int >>= fun m -> int >|= fun n -> (* here *)` `m` and `n` would always be equal, wouldn't they? (yes, I know, for such scenarios it's best to use applicative since generators are uncorrelated, but still, it may happen).
* that being said, is this an actual problem? (this is a real question: I **feel** like it's not, but I may be completely wrong :smile: )

Reply:
In my understanding (so, someone who never write shrinking because of laziness), I prefer to have smaller inputs, as a subparts of the larger... so I slightly prefer your proposition @c-cube

Reply:
I don't think @c-cube 's suggestion would change the size of inputs, only the state `Random` is in when it enters `f` again on shrunk values of `gen`

Reply:
>     * wouldn't this implementation pass a new `st0` copy (i.e. a new, independent state) to both the generator `gen` and to `f a`? I am under the impression that `int >>= fun m -> int >|= fun n -> (* here *)` `m` and `n` would always be equal, wouldn't they? (yes, I know, for such scenarios it's best to use applicative since generators are uncorrelated, but still, it may happen).

Good point. That's probably where splittable generators are useful: split `st`, save a copy of the first half to `gen` (always use  the same copy there) and pass the other half to the function :/

Reply:
If it's ok with you, I am resolving as "May or may not be addressed in https://github.com/c-cube/qcheck/issues/86" :sweat_smile: